utils/trie-compact/src/main.rs

```diff
@@ -0,0 +1,86 @@
+use std::{collections::HashSet, path::PathBuf};
+
+use casper_hashing::Digest;
+use structopt::StructOpt;
+
+use retrieve_state::{
+    storage,
+    storage::{create_storage, normalize_path},
+};
+
+use tracing::info;
+use trie_compact::copy_state_root;
+
+#[derive(Debug, StructOpt)]
+struct Opts {
+    #[structopt(long, required = true, default_value = retrieve_state::CHAIN_DOWNLOAD_PATH)]
+    storage_path: PathBuf,
+
+    #[structopt(long, required = true, default_value = retrieve_state::LMDB_PATH)]
+    source_lmdb_path: PathBuf,
+
+    #[structopt(long, required = true)]
+    destination_lmdb_path: PathBuf,
+}
+
+#[tokio::main]
+async fn main() -> Result<(), anyhow::Error> {
+    let opts = Opts::from_args();
+    env_logger::init();
+
+    let storage_path = normalize_path(&opts.storage_path)?;
+    let source_lmdb_path = normalize_path(&opts.source_lmdb_path)?;
+    let destination_lmdb_path = normalize_path(&opts.destination_lmdb_path)?;
+
+    if source_lmdb_path == destination_lmdb_path {
+        return Err(anyhow::anyhow!(
+            "Source and destination should not be the same lmdb database file: {}",
+            source_lmdb_path.display()
+        ));
+    }
+
+    let (source_state, _env) = storage::load_execution_engine(
+        source_lmdb_path,
+        retrieve_state::DEFAULT_MAX_DB_SIZE,
+        Digest::default(),
+        true,
+    )?;
+
+    let (destination_state, _env) = storage::create_execution_engine(
+        destination_lmdb_path,
+        retrieve_state::DEFAULT_MAX_DB_SIZE,
+        true,
+    )?;
+
+    // Create a separate lmdb for block/deploy storage at chain_download_path.
+    let storage = create_storage(&storage_path).expect("should create storage");
+
+    let mut block = storage.read_highest_block().unwrap().unwrap();
```

Comment:
Same here - especially useful if we add context to each of the two potential failure points.  And for `read_block_by_height` below.